### ISSUE TITLE
question about using spvnas for object detection head

### ISSUE THREAD

**Issue by AllenPeng0209** [mit-han-lab/e3d#13]
Hi, 
Thanks for your awesome work first.

I'm trying to use spvnas backbone for object detection, and trying to merge torchsparse library to MMdetection3d.
but I got stuck in the last layer, the last layer is to add all the feature back to PointTensor, and use a simple classifier to classify the point cloud.

If I want to use SparseTensor for dense head and output the detection result, the simple idea is to make 3D tensor to 2D BEV, 
and use anchor or anchor free head to generate detection result, but it seems like all the tensor.shape is different from the original mmdetection3d.

If it is possible to provide the detection code on kitti, or give some ideas about how to reimplement the result on paper.

thanks!

**Comment by kentangSJTU** [comment]
Hi,

Yes, your understanding is correct. We haven't merged some detection-related functions (such as `ToBEVConvolution`) to our codebase, and I'll try to integrate them to `torchsparse` in this week. Please stay tuned, thanks!


Best,
Haotian

**Comment by AllenPeng0209** [comment]
Thanks for your reply, will stay tuned !

I also want to know is there any function to convert sparsetensor to (B,C,H,W) 
original implementation on SECOND have spconv.SparseConvTensor(features, indices, spatial_shape, batch_size).dense()  
I try to use this function on sparsetensor, but got the error message below, it seems like spatial_shape is not correct.
![Screenshot from 2020-10-09 19-21-57](https://user-images.githubusercontent.com/26760786/96120237-c7c6d980-0f20-11eb-974d-67161f1da3c5.png)

**Comment by kentangSJTU** [comment]
Hi,

I'm not sure whether I understood the question correctly, it seems that `dense()` function should be converting the SparseTensor to a 5D tensor (dense volumetric representation) instead of a 4D one. If it is the function you are looking for, we currently haven't implemented support for such conversion yet, but we may have a look in the future if it is helpful in some applications.

Best,
Haotian

**Comment by AllenPeng0209** [comment]
Hello, 

I want to know the torchsparse detection-related functions merge schedule, I want to make some speed experience compared with SECOND.

Thanks!

**Comment by kentangSJTU** [comment]
Hi Allen,

Sorry for the delay. The code has just been released. Please have a look, thanks!

Best,
Haotian

**Comment by AllenPeng0209** [comment]
Hi Ken,

Nice!
I have tried ToDenseBEVConvolution for making sparsetensor to BEV, and encounter some troubles.
To init ToDenseBEVConvolution, we need to give in_channels, out_channels, shape
It's easy to understand channels, out_channels, but it seems like shape have different definition under each framework.
I try to use ToDenseBEVConvolution in SPVCNN after x4 stage, and I gave this shape(batch_size, C , W, H) or (W, C,H) , and 
dones't work.

also, I change some tensor divide method to torch.floor_divide to support pytorch1.6

Could you simply describe how to use this function and make sparsetensor to  BCHW 

Thanks a lot!

**Comment by AllenPeng0209** [comment]
it seems like your coord shape is incompatible to orginal design.
sparse_coords = (coords - self.offset).t()[[3] + self.bev_dims].long()
I guess what you trying to get is [0, 2, 3] which is (batch_size, H, W)

Thanks for your answer !

**Comment by AllenPeng0209** [comment]
I also want to know that the detection result show on paper is base on SPVCNN y4 and make it to bev for dense head?

Thanks

**Comment by kentangSJTU** [comment]
Hi Allen,

Really sorry for the delay. I'm involved in other projects recently.

Regarding your questions, we indeed miss one function in `torchsparse` latest release last week which is `SparseCrop`, used to crop the `SparseTensor` given some manually specified coordinate range. I've just added this module to `torchsparse`. Basically the conversion from `SparseTensor` to BEV in our code base is done like:

```python
self.to_bev = nn.Sequential(
        SparseCrop(loc_min=loc_min, loc_max=loc_max),
        ToDenseBEVConvolution(in_channels, cs[0], shape=(loc_max-loc_min)//proposal_stride, offset=loc_min),
        nn.BatchNorm2d(cs[0]),
        nn.ReLU(True),
)
```

The `loc_min` and `loc_max` are manually defined crop range. In our old codebase we calculate them in the following way:

```python
self.PC_AREA_SCOPE = np.array([[-40, 40], [-1, 3], [0, 70.4]], 'float32')
self.GT_AREA_SCOPE = {
    'Car': np.array([[-40, 40], [-1, 3], [0, 70.4]], 'float32'),
    'Cyclist': np.array([[-20, 20], [-1, 3], [0, 48]], 'float32'),
    'Pedestrian': np.array([[-20, 20], [-1, 3], [0, 48]], 'float32')
}
self.LOC_MIN = dict([(c, np.array(np.round((self.GT_AREA_SCOPE[c][:, 0] - self.PC_AREA_SCOPE[:, 0]) / self.VOXEL_SIZE), 'int32')) for c in self.classes])
        self.LOC_MAX = dict([(c, np.array(np.round((self.GT_AREA_SCOPE[c][:, 1] - self.PC_AREA_SCOPE[:, 0]) / self.VOXEL_SIZE), 'int32')) for c in self.classes])
```

Proposal stride = 8 for cars and cyclists, and 4 for pedestrians. This simply indicates that for the car/cyclist class, `SparseTensor` is converted to BEV after 3 downsampling stages, and for pedestrians, the conversion happens after 2 downsampling stages. Note that the SPVCNN for detection is not a U-Net, it consists of just one stemming stage (stride=1 spvconv) and 3 downsampling stages similar to what is used in segmentation SPVCNN. There's no upsampling stages.

Hope the answer can at least partially solve your question. 

Best,
Haotian

**Comment by CurryYuan** [comment]
It seems that the function spcrop miss two arguments loc_min and loc_max.

**Comment by kentangSJTU** [comment]
> It seems that the function spcrop miss two arguments loc_min and loc_max.

Hello,

Actually these are `self.LOC_MIN` and `self.LOC_MAX` mentioned above.

Best,
Haotian

**Comment by AllenPeng0209** [comment]
Hi Haotian, 

Thanks for your answer so much.

I still confused about the coors orders, if I want to use (BCHW) as coors, and put it into Sparsetensor and let spnn.3dconv works with it, is that ok to do that?  (I also modify corresponding ToDenseBEVConvolution coors definition) 

I use batch size 1 to do it, it seems like ok, but when I change to batchsize 2, its not working.

I try to merge torchsparse lib to mmdetection3d, and the coors order makes me really confused.

thanks for your reply agian !

**Comment by AllenPeng0209** [comment]
update, i change coors to [HWCB] and assign z dims = 2, it works, and reach to original second mAP.

 now I'm trying to use SPVconv to replace original spconv.

Here is one sample of SPVconv without UNet form, I want to make sure that my understanding from your paper is correct? 

Thanks
`
        #start with SparseTensor
        z = PointTensor(x.F, x.C.float())
        x0 = point_to_voxel(x, z) 
        x0 = self.stem(x0)
        #after stem, we make voxel to point, and use points from previous layer with pointnet and cat these feature to current points feature, then back to voxel.
        z0 = voxel_to_point(x0, z)
        z0.F = z0.F + self.point_transforms[0](z.F)
        x0 = point_to_voxel(x, z)
        # start another spvconv 
        x1 = self.stage1(x0)
`

**Comment by kentangSJTU** [comment]
> update, I change coors to [HWCB] and assign z dims = 2, it works, and reach to original second mAP.
> 
> now I'm trying to use SPVconv to replace the original spconv.
> 
> Here is one sample of SPVconv without UNet form, I want to make sure that my understanding from your paper is correct?
> 
> Thanks
> 
> ` #start with SparseTenso`
> ` z = PointTensor(x.F, x.C.float())`
> ` x0 = point_to_voxel(x, z)`
> ` x0 = self.stem(x0)`
> #after stem, we make voxel to point, and use points from previous layer with pointnet and cat these feature to current points feature, then back to voxel.
> ` z0 = voxel_to_point(x0, z)`
> ` z0.F = z0.F + self.point_transforms[0](z.F)`
> ` x0 = point_to_voxel(x, z)`
> # start another spvconv
> ` x1 = self.stage1(x0) ...`

Hello,

I think this is correct. 

Best,
Haotian

**Comment by jialeli1** [comment]
> update, I change coors to [HWCB] and assign z dims = 2, it works, and reach to original second mAP.
> 
> now I'm trying to use SPVconv to replace the original spconv.
> 
> Here is one sample of SPVconv without UNet form, I want to make sure that my understanding from your paper is correct?
> 
> Thanks
> 
> ` #start with SparseTenso`
> ` z = PointTensor(x.F, x.C.float())`
> ` x0 = point_to_voxel(x, z)`
> ` x0 = self.stem(x0)`
> #after stem, we make voxel to point, and use points from previous layer with pointnet and cat these feature to current points feature, then back to voxel.
> ` z0 = voxel_to_point(x0, z)`
> ` z0.F = z0.F + self.point_transforms[0](z.F)`
> ` x0 = point_to_voxel(x, z)`
> # start another spvconv
> ` x1 = self.stage1(x0) ...`

Hi, bro. Compared to SECOND, the only difference is that the z0 = voxel_to_point(x0, z) is added to the raw points again, and have you found it to be effective?

**Comment by AllenPeng0209** [comment]
@jialeli1 
I did the experiment on kitti, however the Kitti dataset is too small, the loss in the last epoch is smaller than the original version, but it's hard to tell that the results metric is better.

Is that possible to release the detection code on kitti?

@kentangSJTU
